.github/workflows/pipeline.yml

```diff
@@ -0,0 +1,130 @@
+name: Run sytest
+on:
+  push:
+    branches: ["develop", "release-*"]
+  pull_request:
+
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+
+jobs:
+  synapse:
+    runs-on: ubuntu-latest
+
+    strategy:
+      fail-fast: false
+      matrix:
+        include:
+          - sytest-tag: bionic
+
+          - sytest-tag: bionic
+            postgres: postgres
+
+          - sytest-tag: bionic
+            postgres: postgres
+            workers: workers
+
+          - sytest-tag: testing
+            postgres: postgres
+
+          - sytest-tag: testing
+            postgres: postgres
+            workers: workers
+
+    container:
+      image: matrixdotorg/sytest-synapse:${{ matrix.sytest-tag }}
+      volumes:
+        # bootstrap.sh expects the sytest source available at /sytest.
+        # TODO Buildkite mounted /sytest as readonly. Can we do this on GHA? Do we need it?
+        - ${{ github.workspace }}/sytest:/sytest
+        # synapse_sytest.sh expects a synapse checkout at /src
+        - ${{ github.workspace }}/synapse:/src
+      env:
+        POSTGRES: ${{ matrix.postgres && 1 }}
```

Comment:
Is `&& 1` a trick to say that this should be enabled by default?

Why don't we do the same for dendrite's runs?

Reply:
Ahh, I inherited this from synapse's tests.yml too, [see here](https://github.com/matrix-org/synapse/blob/c12b5577f22ee587b60ad7b65e88322ce1d86b7b/.github/workflows/tests.yml#L195-L200).

I think we want the parameters in the `matrix` to have wordy values so they appear nicely in the workflow's output. But some of the BuildKite config used to use config like

```yaml
    env:
      POSTGRES: "1"
      WORKERS: "1"
      BLACKLIST: "synapse-blacklist-with-workers"
```
So I think `&& 1` was a way to ensure we provided the same environment as we were using on buildkite. With that said, synapse_sytest.sh only ever checks `[ -n "$POSTGRES" ]`, i.e. that $POSTGRES is a nonempty string. So maybe `&& 1` wasn't necessary.

Aside: not fully sure how GHA's expressions work, but [there's something written here](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#operators). I think `a && b` evaluates to `b` if both `a` and `b` are truthy, and `false` otherwise.

> Why don't we do the same for dendrite's runs?

Nothing stopping us.  But I realised that dendrite's test cases break down in to two orthogonal factors, whereas
synapse's don't decompose so nicely (`workers` requires `postgres`, I think). This meant that I could make the `matrix` config simpler: see dendrite's lines 88--89 versus synapse's 18--33.

Hmm. With that said, it looks to me that `dendrite_sytest.sh` only ever uses postgres. Maybe the BuildKite config I've been adapting was mistaken?

Reply:
> Hmm. With that said, it looks to me that dendrite_sytest.sh only ever uses postgres. Maybe the BuildKite config I've been adapting was mistaken?

The script always sets up postgres. Some other bits of perl then determine if we're going to actually use postgres. Need to ensure $POSTGRES is non-empty iff we want to use SQLite.

Reply:
Hmm, not sure if I understand. If `POSTGRES` is empty, then we set it to true/1 instead...?

Is the same required for `BLACKLIST` and `API`?

Reply:
Hmmm. I think I wrote out a larger comment but that seems to have disappeared.

I think it might be helpful to work through the details from scratch.

# What the vars actually do
## `POSTGRES`

`POSTGRES` is handled differently for synapse and dendrite.
* In `synapse_sytest.sh` we only read `POSTGRES` in tests `[ -n "$POSTGRES" ]`. We need `POSTGRES` to be empty or not set for SQLite, and nonempty to use postgres.
* In `Dendrite.pm` we test if `POSTGRES` is not defined or equal to `'0'`, and use SQLlite otherwise.

Taking both behaviours into account: 
* to use postgres, set `POSTGRES` to any non-empty string other than `'0'`
* to use sqllite, don't define `POSTGRES`

## `API`
Only for dendrite. If `$ENV{'API'} == '1'` in `Dendrite.pm`, we'll pass `-api` to dendrite. Not sure how that will handle the var being undefined.

## `BLACKLIST`
Only for synapse. If `"$BLACKLIST"` is empty (`test -z`) then we set `BLACKLIST=sytest-blacklist`. It ends up getting passed to `run-tests.pl` via `-B`.

# GitHub Actions' expression language

The language is typed, but presumably `${{ expr }}` converts `expr` to a string before substituting it into the yaml.

Reply:
I found the best way to experiment was to run a dummy action here: https://github.com/DMRobertson/gha-test/

As far as I can tell, `a && b` returns `a` if `a` is falsey, otherwise `b`. No type conversion done on the value that gets returned.

The two cases then:

*  when `postgres` isn't included in the strategy matrix
    - `matrix.postgres` is `null`
    - `null && 1` evaluates to null.
    - `${{ null }}` subsitutes the text `null` into the yaml file.
    - we end up with a yaml KV pair `POSTGRES: null`
    - GHA interprets this as "don't set `POSTGRES`"

* when `postgres: postgres` is in the strategy matrix
  - `matrix.postgres` is `'postgres'`
  - `'postgres && 1` is `1`
  - `${{ 1 }}` substitutes the text `null` into the yaml file
  - yaml KV `POSTGRES: 1`
  - GHA sets the `POSTGRES` env var to the string `1`

I'm about to add a commit which confirms this by dumping the environment before we call the bootstrap script.